app/driver/i2c_master.c

```diff
@@ -77,86 +66,170 @@ i2c_master_setDC(uint8 SDA, uint8 SCL)
     } else {
         I2C_MASTER_SDA_HIGH_SCL_HIGH();
     }
-    if(1 == SCL) { //clock stretching
+    if(1 == SCL) {
         do {
-        } while(0 == GPIO_INPUT_GET(GPIO_ID_PIN(I2C_MASTER_SCL_GPIO)));
+            sclLevel = GPIO_INPUT_GET(GPIO_ID_PIN(I2C_MASTER_SCL_GPIO));
+        } while(sclLevel == 0);
     }
+    i2c_master_wait(5);
+#else
+    uint32 cycles_start;
+    uint32 cycles_curr;
+    uint32 this_SDA_SCL_set_mask = (SDA << i2c[id].pinSDA) | (SCL << i2c[id].pinSCL);
+    uint32 this_SDA_SCL_clear_mask = i2c[id].pin_SDA_SCL_mask ^ this_SDA_SCL_set_mask;
+    i2c[id].m_nLastSDA = SDA;
+    i2c[id].m_nLastSCL = SCL;
+
+    if(i2c[id].cycles_delay > 0){//
```

Comment:
This asm fragment  and loop should be hoisted out of inline into a `static inline` function with a meaningful name and a proper explanation of what you are doing.  The compiler will end up inlining this in the code, so it doesn't have a runtime impact doing this. So you'd endup with
```C
  if(i2c[id].cycles_delay > 0) {
    wait_cycles(i2c[id].cycles_delay);
  }
```
as the inline element

Reply:
I agree, looks cleaner. Already done.